kernel/src/hil/flash.rs

```diff
@@ -125,3 +127,36 @@ pub trait Client<F: Flash> {
     /// Flash erase complete.
     fn erase_complete(&self, error: Error);
 }
+
+pub trait FlashInfo {
+    fn flash_size(&self) -> u32;
+    fn num_pages(&self) -> u32;
+    fn page_size(&self) -> u32;
+    /// Return 0 if flash does not support locking
+    fn num_lock_units(&self) -> u32;
+    /// Return 0 if flash does not support locking
+    fn lock_unit_size(&self) -> u32;
+    /// Return 0 if flash does not support locking
+    fn pages_per_lock_unit(&self) -> u32;
+    fn page_to_lock_unit(&self, page: u32) -> u32;
+}
+
+pub trait FlashLocking {
+    fn lock_unit(&self, page: u32);
+    fn unlock_unit(&self, page: u32);
+    // Locks/unlocks [first,last]
+    fn lock_units(&self, first: u32, last: u32);
+    fn unlock_units(&self, first: u32, last: u32);
+}
+
+pub trait FlashLayout {
```

Comment:
These are separate Traits. It could be that not every chip implements all of them. E.g., FlashLayout only makes sense for the on-chip flash. That being said, I think your point is well taken, the name is too general given what it means. I'd suggest "CodeLayout"? or "SoftwareLayout"?